### PR TITLE
Add Kraken Spot exchange

### PR DESCRIPTION
**Pull request by artememelin**
### Pull request checklist

- [x] Did you bump the project version?
- [x] Did you add a description to the Pull Request?
- [ ] Did you add new tests?
- [x] Did you add reviewers?

### CODE REVIEW

File: src/Kraken/Spot/API/Asset.jl [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    altname::String
+    decimals::Int64
+    display_decimals::Int64
+    collateral_value::Maybe{Float64}
+    status::AssetStatus
+end
+
+function Serde.SerQuery.ser_type(::Type{AssetQuery}, x::Vector{String})::String
+    return join(x, ",")
+end
```
Comment (yshadrin) [inline src/Kraken/Spot/API/Asset.jl:37]: Method must be after the declaration of the `AssetQuery` structure

File: src/Kraken/Spot/API/AssetPair.jl [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+struct FeeData <: KrakenData
+    fee_volume_currency::String
+    fees::Vector{Fee}
+    fees_maker::Vector{Fee}
+end
+
+struct LeverageData <: KrakenData
+    leverage_buy::Maybe{Vector{Int64}}
+    leverage_sell::Maybe{Vector{Int64}}
+end
```
Comment (yshadrin) [inline src/Kraken/Spot/API/AssetPair.jl:64]: I suggest making the following renaming

- `AssetPairData` -> `AssetPairInfoData`
- `FeeData` -> `AssetPairFeeData`
- `LeverageData` -> `AssetPairLeverageData`

File: src/Kraken/Spot/API/Candle.jl [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    x == h1  && return 60
+    x == h4  && return 240
+    x == d1  && return 1440
+    x == w1  && return 10080
+    x == d15 && return 21600
+end
+
+function Serde.SerQuery.ser_type(::Type{CandleQuery}, x::DateTime)::Int64
+    return round(Int64, 1000 * datetime2unix(x))
+end
```
Comment (yshadrin) [inline src/Kraken/Spot/API/Candle.jl:37]: Why is the method `set_type` additionally defined here for the `CandleQuery` type? If in Utils.jl already has an entry
```
function Serde.SerQuery.ser_type(::Type{<:KrakenCommonQuery}, dt::D)::Int64 where {D<:DateTime}
    return round(Int64, 1000 * datetime2unix(dt))
end
```

File: src/Kraken/Spot/API/DepositLog.jl [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+@enum StatusProp begin
+    cancel_pending          # cancelation requested
+    canceled                # canceled
+    cancel_denied           # cancelation requested but was denied
+    _return                 # a return transaction initiated by Kraken; it cannot be canceled
+    onhold                  # withdrawal is on hold pending review
+end
+
+function Serde.deser(::Type{DepositLogQuery}, ::Type{StatusProp}, x::String)::StatusProp
```
Comment (yshadrin) [inline src/Kraken/Spot/API/DepositLog.jl:30]: where `StatusProp` is used in the `DepositLogQuery` structure?

File: src/Kraken/Spot/API/AssetPair.jl [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+struct AssetPairFeeData <: KrakenData
+    fee_volume_currency::String
+    fees::Vector{Fee}
+    fees_maker::Vector{Fee}
+end
+
+struct AssetPairLeverageData <: KrakenData
+    leverage_buy::Maybe{Vector{Int64}}
+    leverage_sell::Maybe{Vector{Int64}}
+end
```
Comment (dmitrii-doronin) [inline src/Kraken/Spot/API/AssetPair.jl:64]: Can they both be nothing? Maybe they return empty arrays

File: src/Kraken/Spot/API/Candle.jl [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    time::Maybe{NanoDate}
+    open::Maybe{Float64}
+    high::Maybe{Float64}
+    low::Maybe{Float64}
+    close::Maybe{Float64}
+    vwap::Maybe{Float64}
+    volume::Maybe{Float64}
+    count::Maybe{Int64}
+    last::Maybe{NanoDate}
+end
```
Comment (dmitrii-doronin) [inline src/Kraken/Spot/API/Candle.jl:45]: It can't all be maybe. At least time should always deserialise

File: src/Kraken/Spot/API/DepositLog.jl [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    refid::String
+    status::String
+    status_prop::Maybe{StatusProp}
+    time::NanoDate
+    txid::String
+end
+
+function Serde.deser(::Type{DepositLogData}, ::Type{StatusProp}, x::String)::StatusProp
+    x == "return" && return _return
+    x == "onhold" && return onhold
```
Comment (dmitrii-doronin) [inline src/Kraken/Spot/API/DepositLog.jl:46]: Why "onhold" though?

File: src/Kraken/Spot/API/Ticker.jl [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    o::Maybe{Float64}
+end
+
+function Serde.deser(::Type{Level}, x::Vector{Any})::Level
+    return Level(
+        tryparse(Float64, x[1]),
+        tryparse(Float64, x[2]),
+        tryparse(Float64, x[3]),
+    )
+end
```
Comment (dmitrii-doronin) [inline src/Kraken/Spot/API/Ticker.jl:42]: Seems unnecessary.

File: src/Kraken/Spot/API/Ticker.jl [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    a::Maybe{Level}
+    b::Maybe{Level}
+    c::Maybe{Vector{Float64}}
+    v::Maybe{Vector{Float64}}
+    p::Maybe{Vector{Float64}}
+    t::Maybe{Vector{Int64}}
+    l::Maybe{Vector{Float64}}
+    h::Maybe{Vector{Float64}}
+    o::Maybe{Float64}
+end
```
Comment (dmitrii-doronin) [inline src/Kraken/Spot/API/Ticker.jl:34]: Something must be always present

File: src/Kraken/Spot/API/AssetPair.jl [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    pair::Maybe{String} = nothing
+    info::AssetPairInfo = info
+end
+
+struct Fee <: KrakenData
+    volume::Int64
+    percent::Float64
+end
+
+struct AssetPairData <: KrakenData
```
Comment (yshadrin) [inline src/Kraken/Spot/API/AssetPair.jl:26]:
```suggestion
struct AssetPairInfoData <: KrakenData
```
for uniformity

File: src/Kraken/Spot/API/DepositLog.jl [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+end
+
+"""
+    deposit_log(client::KrakenClient, query::DepositLogQuery)
+    deposit_log(client::KrakenClient = Kraken.Spot.public_client; kw...)
```
Comment (yshadrin) [inline src/Kraken/Spot/API/DepositLog.jl:51]: private method
```suggestion
    deposit_log(client::KrakenClient; kw...)
```

Comment (yshadrin) [reply src/Kraken/Spot/API/DepositLog.jl:51]: and fix it in all cases

File: src/Kraken/Utils.jl [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    return NanoDate(DateTime(x, "yyyy-mm-ddTHH:MM:SS.ssszzzz"))
+end
+
+function Serde.ser_ignore_field(::Type{<:KrakenCommonQuery}, ::Val{:signature})::Bool
+    return true
+end
+
+function Serde.SerQuery.ser_type(::Type{<:KrakenCommonQuery}, dt::D)::Int64 where {D<:DateTime}
+    return round(Int64, 1000 * datetime2unix(dt))
+end
```
Comment (yshadrin) [inline src/Kraken/Utils.jl]:
```suggestion
function Serde.SerQuery.ser_type(::Type{<:KrakenCommonQuery}, x::D)::Int64 where {D<:DateTime}
    return round(Int64, 1000 * datetime2unix(x))
end
```

File: examples/Kraken/Spot.jl [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    interval = CryptoAPIs.Kraken.Spot.Candle.h1,
+    since = now(UTC) - Hour(1),
+)
+
+CryptoAPIs.Kraken.Spot.order_book(; pair = "XBTUSD", count = 10)
+
+CryptoAPIs.Kraken.Spot.ticker()
+CryptoAPIs.Kraken.Spot.ticker(; pair = "XBTUSD")
+
+kraken_client = CryptoAPIs.Kraken.KrakenClient(;
```
Comment (yshadrin) [inline examples/Kraken/Spot.jl:24]:
```suggestion
kraken_client = KrakenClient(;
```

File: src/Kraken/Kraken.jl [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+Exception thrown when an API method fails with code `T`.
+
+## Required fields
+- `error::Vector{String}`: Error values.
+"""
+struct KrakenAPIError{T} <: AbstractAPIsError
+    error::Vector{String}
+
+    function APIsError(error::Vector{String})
```
Comment (yshadrin) [inline src/Kraken/Kraken.jl:72]:
```suggestion
    function KrakenAPIError(error::Vector{String})
```

File: src/Kraken/Kraken.jl [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    return query
+end
+
+function CryptoAPIs.request_body(::Q)::String where {Q<:KrakenCommonQuery}
+    return ""
+end
+
+function CryptoAPIs.request_query(query::Q)::String where {Q<:KrakenCommonQuery}
+    return Serde.to_query(query)
+end
```
Comment (yshadrin) [inline src/Kraken/Kraken.jl:107]:
```suggestion
function CryptoAPIs.request_body(::Q)::String where {Q<:KrakenPublicQuery}
    return ""
end

function CryptoAPIs.request_body(query::Q)::String where {Q<:KrakenPrivateQuery}
    return Serde.to_query(query)
end

function CryptoAPIs.request_query(query::Q)::String where {Q<:KrakenPublicQuery}
    return Serde.to_query(query)
end

function CryptoAPIs.request_query(::Q)::String where {Q<:KrakenPrivateQuery}
    return ""
end
```
it is necessary to separately describe the methods for KrakenPublicQuery and KrakenPrivateQuery
this is important for the API to work correctly

File: src/Kraken/Utils.jl [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+function Serde.deser(::Type{<:KrakenData}, ::Type{<:Maybe{NanoDate}}, x::AbstractString)::NanoDate
+    return NanoDate(DateTime(x, "yyyy-mm-ddTHH:MM:SS.ssszzzz"))
+end
+
+function Serde.ser_ignore_field(::Type{<:KrakenCommonQuery}, ::Val{:signature})::Bool
+    return true
+end
+
+function Serde.SerQuery.ser_type(::Type{<:KrakenCommonQuery}, x::D)::Int64 where {D<:DateTime}
+    return round(Int64, datetime2unix(x))
```
Comment (yshadrin) [inline src/Kraken/Utils.jl:16]:
```suggestion
    return round(Int64, 1000 * datetime2unix(x))
```

File: src/Kraken/Utils.jl [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    return unixnanos2nanodate(x * 1e9)
+end
+
+function Serde.deser(::Type{<:KrakenData}, ::Type{<:Maybe{NanoDate}}, x::AbstractString)::NanoDate
+    return NanoDate(DateTime(x, "yyyy-mm-ddTHH:MM:SS.ssszzzz"))
+end
+
+function Serde.ser_ignore_field(::Type{<:KrakenCommonQuery}, ::Val{:signature})::Bool
+    return true
+end
```
Comment (yshadrin) [inline src/Kraken/Utils.jl:13]: `signature` is necessary for private request

**Pull request merged**